skbio/metadata/_interval.py

```diff
@@ -0,0 +1,465 @@
+        If this isn't specified, then all of the boundaries are true.
+    metadata : dict
+        Dictionary of attributes storing information of the feature
+        such as `strand`, `gene_name` or `product`.
+
+    See Also
+    --------
+    skbio.metadata.IntervalMetadata
+    '''
+    def __init__(self, interval_metadata, intervals,
+                 boundaries=None, metadata=None):
+
+        ivs = list(intervals)
+
+        # Intervals
+        self._interval_metadata = interval_metadata
+        # http://stackoverflow.com/a/6422754/1167475
+        # Used to sort boundaries later
+
+        indices = [i[0] for i in sorted(enumerate(ivs),
+                                        key=operator.itemgetter(1))]
+        self.intervals = sorted(ivs)
+
+        # Boundaries
+        if boundaries is not None:
+            bds = list(boundaries)
+            if len(bds) != len(ivs):
+                msg = ('The number of boundaries (%d) needs '
+                       'to be equal to the number of intervals (%d).')
+                raise ValueError(msg % (len(bds), len(ivs)))
+            self.boundaries = [bds[i] for i in indices]
+        else:
+            self.boundaries = [(True, True)] * len(ivs)
+
+        # Metadata
+        if metadata is not None:
+            self._metadata = metadata
+        else:
+            self._metadata = {}
+
+        if interval_metadata is not None:
+            self._add_interval()
+
+    def _add_interval(self):
+        """ Add an interval object to the IntervalTree within
+            IntervalMetadata."""
+        # Add directly to the tree.  So no need for _is_stale_tree
+        for loc in self.intervals:
+            start, end = loc
+            self._interval_metadata._intervals.add(start, end, self)
+        self._interval_metadata._metadata.append(self)
+
+    # This is required for creating unique sets of intervals
+    def _hash(self):
+        return hash(tuple(sorted(self.metadata.items()) +
+                          self.intervals +
+                          self.boundaries))
+
+    def _cmp(self, other):
+        """ Comparison operator required for sorting intervals.
+
+        Note that this method does not check for equality.
+        """
+        return self.intervals < other.intervals
+
+    @experimental(as_of='0.4.2-dev')
+    def __eq__(self, other):
+        return ((self.metadata == other.metadata) and
+                (self.intervals == other.intervals) and
+                (self.boundaries == other.boundaries))
+
+    @experimental(as_of='0.4.2-dev')
+    def __ne__(self, other):
+        return not self.__eq__(other)
+
+    @experimental(as_of='0.4.2-dev')
+    def __repr__(self):
+        # need pformat to print out the dictionary
+        # in a consistent manner
+        classname = self.__class__.__name__
+        num_intervals = len(self.intervals)
+        num_keys = len(self.metadata)
+        sts, ends = zip(*self.intervals)
```

Comment:
This will fail if ``self.intervals`` is empty. 

```python
>>> sts, ends = zip(*[])
ValueError: not enough values to unpack (expected 2, got 0)
```

Please fix and add a test case to cover this.